{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Running a quality control (QC) analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Introduction\n",
    "\n",
    "One of the most important steps in any RNA-Seq analysis is to quality control (QC) check your data. By default, DEAGO will run both a QC and a DE analysis. You can ask DEAGO to build a quick QC report by using the `--qc` option.  This is a great way to get a first look at your data to try and identify any issues e.g. batch effects and outliers.\n",
    "\n",
    "The objectives of this part of the tutorial are:\n",
    "\n",
    "  * run a QC analysis with DEAGO\n",
    "  * interpret the output QC report from DEAGO\n",
    "\n",
    "\n",
    "### Input files\n",
    "\n",
    "We will need to give DEAGO two bits of information:\n",
    "\n",
    "  * *the name/location of the directory containing our gene count files (counts)*\n",
    "  \n",
    "  \n",
    "  * *the name/location of our sample/condition mapping file (targets.txt)*\n",
    "  \n",
    "  \n",
    "### Running a QC analysis with DEAGO\n",
    "\n",
    "To run a quick, QC analysis with DEAGO the command would be:\n",
    "\n",
    "```\n",
    "deago -c <counts_directory> -t <targets file> --qc\n",
    "```\n",
    "\n",
    "As our count files were generated by featureCounts for this tutorial, we need to also tell DEAGO the count format with the `--count_type` option:\n",
    "\n",
    "```\n",
    "deago -c <counts_directory> -t <targets file> --count_type featurecounts --qc\n",
    "```\n",
    "\n",
    "### QC output files\n",
    "\n",
    "Once your QC analysis has finished, you should see several new files:\n",
    "\n",
    "  * **`deago.config`**  \n",
    "  _config file with key/value parameters defining the analysis_  \n",
    "\n",
    "\n",
    "  * **`deago.rlog`**  \n",
    "  _log of the R output generated when converting the R markdown to HTML_  \n",
    "\n",
    "\n",
    "  * **`deago_markdown.Rmd`**  \n",
    "  _R markdown used to run the analysis_  \n",
    "\n",
    "\n",
    "  * **`deago_markdown.html`**  \n",
    "  _HTML report generated from the R markdown_   \n",
    "  \n",
    "See our [user guide](https://github.com/sanger-pathogens/Bio-Deago/blob/master/user_guide/Output-files.ipynb) more information about the output files from DEAGO.\n",
    "\n",
    "### QC analysis report\n",
    "\n",
    "The output file we're interested in is **`deago_markdown.html`** which is your QC analysis report.  Go ahead and open it in a web browser (e.g. Chrome, Firefox, IE, Safari...). You can do this by going to \"File -> Open\" in the top navigation or (if you have Firefox installed, use the command:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "firefox deago_markdown.html"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "All DEAGO reports have a sidebar on the left so you can quickly navigate the report."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "![Navigation panel](images/deago_navigation.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If you click on **`Pipeline configuration`** you will see that the report contains the commands used to run the analysis (grey boxes) and their output. The **`Pipeline configuration`** section shows the parameters that were used for the analysis. This can be useful for finding input/output data, troubleshooting and debugging."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "![Pipeline configuration](images/deago_parameters.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To take a look at the QC plots, you can click on **`QC Plots`** in the left-hand panel.\n",
    "\n",
    "First, check to see if there were any problems with the sequencing. The _Total read counts per sample_ plot does what it says on the tin and shows you the number of reads overlapping features in each sample. If one sample has a particularly low count compared to the others, it may indicate an issue and so you should take a closer look at that sample in some of the other QC plots."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "![Total read counts per sample plot](images/reads_per_sample.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The principal component analysis (PCA) and sample-to-sample distances plot are indicators of variation and will show how well the samples cluster together. These are a quick way of spotting outliers and potential batch effects.\n",
    "\n",
    "In the PCA plot, samples are coloured by condition and the sample labels are a combination of the sample condition and the replicate number.  Here you will see whether your samples cluster reasonably well with distinct groups for each condition: **wt_ctrl**, **wt_il22** and **ko_ctrl**/**ko_il22**."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "![PCA plot](images/pca.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To compliment the PCA plot, we also have a scree plot which is a histogram of the percentage contribution of each of the principal components (PC). The points indicate the cumulative total and the lines represent broad cutoffs of 70% and 90%.  Ideally, in a two-factor analysis we'd hope to see the cumulative total of PC1 and PC2 reaching 70%, although this may not always be the case."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "![PCA scree plot](images/pca_scree.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There is also a sample-to-sample distances plot shaded by the distance (or variability) between samples. The darker the box, the more similar the samples (or the smaller the distance between them)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "![Sample distances plot](images/sample_distances.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercise 3"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**First, let's make sure we're in the `data` directory.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "cd data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Each DEAGO analysis should be self-contained, so let's create a new directory for our QC analysis.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "mkdir qc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "cd qc"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**As this is the first time we're running `deago` in this tutorial, let's take a look at the usage.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "deago -h"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here's a brief explanation of the options we want to use and what they mean:\n",
    "\n",
    "  * **`--build_config`**  \n",
    "  _tells DEAGO that we want to build a new config file using the command line parameters_ \n",
    "\n",
    "\n",
    "  * **`-c`**  \n",
    "  _tells DEAGO the location of the folder containing our count files_\n",
    "\n",
    "\n",
    "  * **`-t`**  \n",
    "  _tells DEAGO the location of our sample/condition mapping file_\n",
    "\n",
    "\n",
    "  * **`--count_type`**  \n",
    "  _tells DEAGO the format of the count data (e.g. featurecounts) setting the values for **count&#95;column**, **skip&#95;lines**, **gene&#95;ids** and **count&#95;delim**_\n",
    "\n",
    "\n",
    "  * **`--qc`**  \n",
    "  _tells DEAGO to only run the QC analysis_"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Now, let's get our QC report.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "deago --build_config -c ../counts -t ../targets.txt \\\n",
    "    --count_type featurecounts --qc"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "_Note: because we are running the analysis from our new `qc` directory, we need to us `../` to say that our directory is one level above where we are now._"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Questions\n",
    "\n",
    "**Q1: Do all the samples have similar total read counts?**\n",
    "\n",
    "**Q2: Look at the PCA plot. How many clusters have the samples grouped into?**\n",
    "\n",
    "**Q3: Do you notice anything in the PCA and sample-to-sample distances plot that you might want to look closer at?**  \n",
    "_Hint: look at the groupings in both plots, are there sub-groupings, do they relate to anything other than the condition..._\n",
    "\n",
    "**Q4: What does the `--keep_images` option do?**  \n",
    "_Hint: look at the DEAGO usage with `deago -h`_"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## What's next?\n",
    "\n",
    "If you want a recap of input file preparation, head back to [preparing input data](input-data.ipynb).\n",
    "\n",
    "Otherwise, let's continue on to [running a differential expression (DE) analysis](differential-expression.ipynb)."
   ]
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
