{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Running Roary\n",
    "\n",
    "At this stage you should have three GFF files generated by Prokka, each in its own directory. Provided your QC looked alright, you are now ready to run Roary to generate the pan genome. \n",
    "\n",
    "We are going to run Roary twice, first with the default settings, and then using MAFFT to generate a core gene alignment. For both of these runs we will want all the annotation files in the same directory, so lets take a copy of them to our current directory:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "cp annotated_sample*/*.gff ."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Run Roary with default settings\n",
    "Running Roary with the default settings is very straightforward. All you need to do is to run `roary *.gff` and it will create a pan genome using all GFF files in the current directory. We want to run Roary twice with different settings, so in order to keep track of our output files from each run we will also specify an output directory where Roary should put the results. Give the following command a go:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "roary -f output_no_alignment *.gff"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This will run for a minute or two.\n",
    "\n",
    "We will have a closer look at the results in the next section, so for now let us just see that there are some output files in the directroy we asked Roary to create:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "ls -l output_no_alignment"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Run Roary with MAFFT\n",
    "To be able to create pretty trees and cool visualisations, we want to genereate a multi-FASTA alignment of the core genes. To do this, we will now run Roary again, but this time with some more options.\n",
    "\n",
    "|Option |Description                                    |\n",
    "|------ |-----------                                    |\n",
    "|-e     |Create a multiFASTA alignment of the core genes|\n",
    "|--mafft|Use with -e to use MAFFT instead of PRANK      |\n",
    "|-p     |Number of threads to use                       |\n",
    "  \n",
    "By default, Roary will use PRANK when the -e option is speified. It is accurate but slow. MAFFT is less accurate but very fast so we are going to use this instead by specifying the --mafft option. To further speed things up, we are going to use 8 threads (the -p option). For all usage options, you can have a look at the [Roary website](https://sanger-pathogens.github.io/Roary/)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "roary -f output_with_alignment -e --mafft -p 8 *.gff"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This will take a bit longer to run than the previous command, about 5 minutes. Once finished you should have a directory called output_with_alignment containing the output files, this time including a core_gene_alignment.aln file. Just quickly check that this is the case and then we will head over to the next section: [Exploring the results](results.ipynb)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "ls -l output_with_alignment"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Check your understanding\n",
    "**Q7: Why do we want to run Roary with MAFFT?**  \n",
    "a) Because it's quicker than to run Roary without the -e option  \n",
    "b) To get more accurate results  \n",
    "c) To generate a core gene alignment  \n",
    "  \n",
    "**Q8: Why do we use the -p otion?**  \n",
    "a) We have to when we use MAFFT  \n",
    "b) To speed up the run  \n",
    "c) To get a pretty tree  \n",
    "  \n",
    "The answers to these questions can be found [here](answers.ipynb).  \n",
    "\n",
    "You can also revisit the [previous section](qc.ipynb), or go back to the [index page](index.ipynb)."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
